ethers-core/src/abi/packed.rs

```diff
@@ -0,0 +1,423 @@
+/// # use ethers_core::types::{Address, H256};
+/// # use ethers_core::utils::keccak256;
+///
+/// let factory: Address = "0x5C69bEe701ef814a2B6a3EDD4B1652CB9cc5aA6f".parse()?;
+///
+/// let token_a: Address = "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48".parse()?;
+/// let token_b: Address = "0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2".parse()?;
+/// let encoded = encode_packed(&[Token::Address(token_a), Token::Address(token_b)])?;
+/// let salt = keccak256(encoded);
+///
+/// let init_code_hash: H256 = "0x96e8ac4277198ff8b6f785478aa9a39f403cb768dd02cbee326c3e7da348845f".parse()?;
+///
+/// let pair = ethers_core::utils::get_create2_address_from_hash(factory, salt, init_code_hash);
+/// let weth_usdc = "0xB4e16d0168e52d35CaCD2c6185b44281Ec28C9Dc".parse()?;
+/// assert_eq!(pair, weth_usdc);
+///
+/// # Ok::<(), Box<dyn std::error::Error>>(())
+/// ```
+pub fn encode_packed(tokens: &[Token]) -> Result<Vec<u8>, EncodePackedError> {
+    // Get vec capacity and find invalid tokens
+    let mut max = 0;
+    for token in tokens {
+        check(token)?;
+        max += max_encoded_length(token);
+    }
+
+    // Encode the tokens
+    let mut bytes = Vec::with_capacity(max);
+    for token in tokens {
+        encode_token(token, &mut bytes, false);
+    }
+    Ok(bytes)
+}
+
+/// The maximum byte length of the token encoded using packed mode.
+fn max_encoded_length(token: &Token) -> usize {
+    match token {
+        Int(_) | Uint(_) | FixedBytes(_) => 32,
+        Address(_) => 20,
+        Bool(_) => 1,
+        // account for padding
+        Array(vec) | FixedArray(vec) | Tuple(vec) => {
+            vec.iter().map(|token| max_encoded_length(token).max(32)).sum()
+        }
+        Bytes(b) => b.len(),
+        String(s) => s.len(),
+    }
+}
+
+/// Tuples and nested arrays are invalid in packed encoding.
+fn check(token: &Token) -> Result<(), EncodePackedError> {
+    match token {
+        FixedBytes(vec) if vec.len() > 32 => Err(EncodePackedError::InvalidBytesLength),
+
+        Tuple(_) => Err(EncodePackedError::InvalidToken(token.clone())),
+        Array(vec) | FixedArray(vec) => {
+            for t in vec.iter() {
+                if t.is_dynamic() || matches!(t, Array(_)) {
+                    return Err(EncodePackedError::InvalidToken(token.clone()))
+                }
+                check(t)?;
+            }
+            Ok(())
+        }
+
+        _ => Ok(()),
+    }
+}
+
+/// Encodes `token` as bytes into `out`.
+fn encode_token(token: &Token, out: &mut Vec<u8>, in_array: bool) {
+    match token {
+        // Padded to 32 bytes if in_array
+        Address(addr) => {
+            if in_array {
+                out.extend_from_slice(&[0; 12]);
+            }
+            out.extend_from_slice(&addr.0)
+        }
+        Int(n) | Uint(n) => {
+            // TODO: Different (u)int* padding
+            let mut buf = [0; 32];
+            n.to_big_endian(&mut buf);
+            out.extend_from_slice(&buf);
+        }
```

Comment:
By using the same `Token` type as `ethabi`, we cannot differentiate between the different integer type sizes as `Token::(U)Int` stores only the  `U256` number and not its encoded size (since in normal abi-encoding this wouldn't have any effect).
Should this function use a different `Token`-like enum, or is there another way around this?

Reply:
You can find the smallest value for which `n >> BITS` is 0, and mark that as the size?

e.g.
255 >> 8  == 0 --> put it in uint8
whereas 256 >> 8 == 1, putting it in uint16
etc.

Reply:
Right, but how does one then cast the number as something higher than its minimum bits? Like uint256(0) would need to pad all 32 bytes? We can have this as a minor limitation for the time being I guess

Reply:
Implemented this using `U256::bits`. Any 0 padding needs to be manually added with something like this:

```rust
let n = U256::from(...);
let mut buf = [0; 32];
n.to_big_endian(&mut buf);
// uint32 = 4 bytes -> 32 - 4 .. 32
let uint32 = Token::Bytes(buf[28..32]);
```